File: whisper/ZeticMLangeWhisper-iOS/ZeticMLangeWhisper-iOS/WhisperDecoder.swift
import Foundation
import ZeticMLange

class WhisperDecoder {
    private let endToken = 50257
    
    private var model: ZeticMLangeModel
    
    init(_ modelKey: String) {
        model = ZeticMLangeModel(modelKey)!
    }
    
    func process(_ encoderOutput: Data, _ maxLength: Int = 80, temperature: Float = 1.0) -> [Int32] {
        var decoderInputIds = Array(repeating: Int32(0), count: maxLength)
        decoderInputIds[0] = 50258
        decoderInputIds[1] = 50264
        decoderInputIds[2] = 50359
        
        var generatedIds: [Int32] = []
        
        for i in 3..<maxLength {
            let logits = decodeStep(decoderInputIds, encoderOutput)
            
            let logitsSize = logits.count / maxLength
            
            let scaledLogits: [Float]
            if temperature > 0 {
                scaledLogits = logits.map { $0 / temperature }
            } else {
                scaledLogits = logits
            }
            
            let startIndex = logitsSize * (i - 1)
            let endIndex = logitsSize * i
            let currentLogits = Array(scaledLogits[startIndex..<endIndex])
            
            let probs = ProbabilityUtils.softmax(currentLogits)
            
            let nextToken: Int
            if temperature > 0 {
                nextToken = ProbabilityUtils.sampleFromDistribution(probs)
            } else {
                nextToken = ProbabilityUtils.argmax(probs)
            }
            
            if nextToken == endToken {
                break
            }
            
            generatedIds.append(Int32(nextToken))
            
            decoderInputIds[i] = Int32(nextToken)
        }
        
        print(generatedIds.map { "\($0)"}.joined(separator: ","))
        
        return generatedIds
    }
    
    private func decodeStep(_ decoderInputIds: [Int32], _ encoderOutput: Data) -> [Float] {
        do {
            try model.run([
                Data(bytes: decoderInputIds, count: 80 * MemoryLayout<Int32>.size),
                encoderOutput
            ])
        } catch {
            
        }
        
        var array = model.getOutputDataArray()[0].withUnsafeBytes {
            Array($0.bindMemory(to: Float.self))
        }
        let gap: Int = Int((ceilf(51865 / 32) * 32) - 51865)
        for i in 1..<80 {
            array.removeSubrange(51865 * i..<51865 * i + gap)
            array = array + (Array(repeating: Float(0), count: gap))
        }
        return array
    }
}

File: whisper/ZeticMLangeWhisper-Android/app/src/main/java/com/zeticai/zeticmlangewhisper_android/WhisperDecoder.kt
package com.zeticai.zeticmlangewhisper_android

import android.content.Context
import com.zetic.ZeticMLange.ZeticMLangeDataUtils
import com.zetic.ZeticMLange.ZeticMLangeModel
import java.nio.ByteBuffer
import kotlin.time.measureTimedValue

class WhisperDecoder(
    private val startToken: Int,
    private val languageToken: Int,
    private val taskToken: Int,
    private val endToken: Int,
    context: Context,
    modelKey: String,
    private val model: ZeticMLangeModel = ZeticMLangeModel(context, modelKey)
) {
    fun generateTokens(
        encoderOutput: ByteBuffer,
        maxLength: Int = 80,
        temperature: Float = 1.0f
    ): List<Int> {
        // Initialize with special tokens
        var decoderInputIds = LongArray(80).apply {
            set(0, startToken.toLong())
            set(1, languageToken.toLong())
            set(2, taskToken.toLong())
        }

        val generatedIds = mutableListOf<Int>()

        for (i in 3 until maxLength) {
            val logits = measureTimedValue {
                decodeStep(encoderOutput, decoderInputIds)
            } .also {
                println("decode inference : ${it.duration}")
            }.value

            val logitsSize = logits.size / maxLength

            val scaledLogits = if (temperature > 0) {
                FloatArray(logits.size) { j -> logits[j] / temperature }
            } else {
                logits
            }
            val currentLogits = scaledLogits.slice((logitsSize * (i - 1)) until (logitsSize * (i))).toFloatArray()

            val probs = ProbabilityUtils.softmax(currentLogits)

            val nextToken = if (temperature > 0) {
                ProbabilityUtils.sampleFromDistribution(probs)
            } else {
                ProbabilityUtils.argmax(probs)
            }

            if (nextToken == endToken) {
                break
            }

            generatedIds.add(nextToken)

            decoderInputIds[i] = nextToken.toLong()
        }

        return generatedIds
    }

    private fun decodeStep(encoderOutput: ByteBuffer, decoderInputIds: LongArray): FloatArray {
        model.run(arrayOf(
            ZeticMLangeDataUtils.convertLongArrayToByteBuffer(decoderInputIds),
            encoderOutput,
            )
        )

        val outputs = model.outputBuffers
        return ZeticMLangeDataUtils.convertByteBufferToFloatArray(outputs[0])
    }
}